task/control-plane-agent/src/mgs_sidecar/monorail_port_status.rs

```diff
@@ -0,0 +1,171 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use drv_monorail_api::{Monorail, MonorailError};
+use gateway_messages::vsc7448_port_status::{
+    LinkStatus, PacketCount, PhyStatus, PhyType, PortConfig, PortCounters,
+    PortDev, PortMode, PortSerdes, PortStatus, PortStatusError,
+    PortStatusErrorCode, Speed,
+};
+
+pub(super) fn port_status(
+    monorail: &Monorail,
+    port: u32,
+) -> Result<PortStatus, PortStatusError> {
+    let (port_status, counters, phy_status) = monorail
+        .get_port_status(port as u8)
+        .and_then(|port_status| {
+            monorail
+                .get_port_counters(port as u8)
+                .map(|counters| (port_status, counters))
+        })
+        .and_then(|(port_status, counters)| {
+            match monorail.get_phy_status(port as u8) {
+                Ok(phy_status) => Ok((port_status, counters, Some(phy_status))),
+                Err(MonorailError::NoPhy) => Ok((port_status, counters, None)),
+                Err(err) => Err(err),
+            }
+        })
+        .map_err(|err| {
+            let code = match err {
+                MonorailError::UnconfiguredPort => {
+                    PortStatusErrorCode::Unconfigured
+                }
+                _ => PortStatusErrorCode::Other(err as u32),
+            };
+            PortStatusError { port, code }
+        })?;
+    Ok(PortStatus {
+        port,
+        cfg: PortConfigConvert(port_status.cfg).into(),
+        link_status: LinkStatusConvert(port_status.link_up).into(),
+        phy_status: phy_status.map(|s| PhyStatusConvert(s).into()),
+        counters: PortCountersConvert(counters).into(),
+    })
+}
+
+struct PortConfigConvert(drv_monorail_api::PortConfig);
```

Comment:
Vague question: is having conversion structs and implementing `Into` on them cleaner than free functions?

e.g.
```rust
fn convert_port_config(p: drv_monorail_api::PortConfig) -> PortConfig {
    PortConfig {
        mode: convert_port_mode(s.mode),
        dev: (convert_port_dev(s.dev.0), s.dev.1),
        serdes: (convert_port_serdes(s.serdes.0), s.serdes.1),
    }
}
```
(I've got a weak preference for free functions, unless there's a benefit I'm not seeing)

Reply:
No benefit, just a weak preference in the other direction. :man_shrugging: Without orphan rules I'd obviously implement `From` directly, so this feels like "I'm just working around orphan rules to do these conversions"; using free functions seems slightly less obvious to me?

Reply:
Okay, that seems fine